File: src/asaree/services/factor_bindings.py
"""Validation of the bridge between declared factors and a protocol graph."""

from __future__ import annotations

from typing import Any

_MISSING = object()


def _path_value(root: dict[str, Any], dotted_path: str) -> Any:
    value: Any = root
    for part in dotted_path.split("."):
        if not isinstance(value, dict) or part not in value:
            return _MISSING
        value = value[part]
    return value


def unbound_factor_names(design_spec: dict[str, Any] | None, graph: dict[str, Any]) -> list[str]:
    declared = [str(f.get("name") or "").strip() for f in (design_spec or {}).get("factors") or []]
    bound = {
        str(name)
        for node in graph.get("nodes") or []
        for name in ((node.get("data") or {}).get("factor_bindings") or {}).values()
    }
    return [name for name in declared if name and name not in bound]


def validate_factor_bindings(design_spec: dict[str, Any] | None, graph: dict[str, Any]) -> None:
    missing = unbound_factor_names(design_spec, graph)
    if missing:
        labels = ", ".join(repr(name) for name in missing)
        raise ValueError(f"Rebind or remove unbound experimental factor(s): {labels}.")

    factors = {
        str(factor.get("name") or "").strip(): factor
        for factor in (design_spec or {}).get("factors") or []
        if str(factor.get("name") or "").strip()
    }
    for node in graph.get("nodes") or []:
        data = node.get("data") or {}
        label = str(data.get("label") or node.get("type") or node.get("id") or "node")
        for field_path, raw_factor_name in (data.get("factor_bindings") or {}).items():
            factor_name = str(raw_factor_name).strip()
            factor = factors.get(factor_name)
            if factor is None:
                raise ValueError(
                    f"{label!r} field {field_path!r} is bound to undeclared factor {factor_name!r}. "
                    "Remove the stale binding or restore the factor."
                )
            published_value = _path_value(data, str(field_path))
            if published_value is _MISSING:
                raise ValueError(
                    f"{label!r} field {field_path!r}, bound to factor {factor_name!r}, no longer exists on the "
                    "published canvas. Remove the stale binding or restore the field."
                )
            levels = factor.get("levels") or []
            if not any(published_value == level for level in levels):
                raise ValueError(
                    f"{label!r} field {field_path!r} is bound to factor {factor_name!r}, but its published value "
                    "does not match any declared level. Update the factor levels and regenerate the design, or "
                    "remove the binding before publishing or running."
                )


__all__ = ["unbound_factor_names", "validate_factor_bindings"]

File: frontend/src/lib/factorBindings.ts
import type { DesignSpec } from '@/types/experiments'
import type { ProtocolGraph } from '@/types/protocols'

export interface FactorBindingDiscrepancy {
  nodeId: string
  nodeLabel: string
  fieldPath: string
  factorName: string
  reason: 'factor is no longer declared' | 'field no longer exists on the canvas' | 'published value does not match any declared level'
}

const MISSING = Symbol('missing factor-bound field')

function pathValue(root: Record<string, unknown>, dottedPath: string): unknown | typeof MISSING {
  let value: unknown = root
  for (const part of dottedPath.split('.')) {
    if (typeof value !== 'object' || value === null || !(part in value)) return MISSING
    value = (value as Record<string, unknown>)[part]
  }
  return value
}

function valuesEqual(left: unknown, right: unknown): boolean {
  return JSON.stringify(left) === JSON.stringify(right)
}

/** Declared factors without a live canvas field to vary are not runnable. */
export function unboundFactorNames(designSpec: DesignSpec | null, graph: ProtocolGraph | undefined): string[] {
  const declared = designSpec?.factors?.map((factor) => factor.name.trim()).filter(Boolean) ?? []
  const bound = new Set(
    (graph?.nodes ?? []).flatMap((node) => Object.values(node.data.factor_bindings ?? {})),
  )
  return declared.filter((name) => !bound.has(name))
}

/** Factor bindings that would make a cell replace what the canvas currently shows. */
export function factorBindingDiscrepancies(
  designSpec: DesignSpec | null | undefined,
  graph: ProtocolGraph | undefined,
): FactorBindingDiscrepancy[] {
  const factors = new Map((designSpec?.factors ?? []).map((factor) => [factor.name.trim(), factor]))
  const discrepancies: FactorBindingDiscrepancy[] = []

  for (const node of graph?.nodes ?? []) {
    const data = node.data as unknown as Record<string, unknown>
    const nodeLabel = typeof data.label === 'string' && data.label ? data.label : node.type ?? node.id
    const bindings = (data.factor_bindings ?? {}) as Record<string, string>
    for (const [fieldPath, rawFactorName] of Object.entries(bindings)) {
      const factorName = rawFactorName.trim()
      const factor = factors.get(factorName)
      const value = pathValue(data, fieldPath)
      const reason = !factor
        ? 'factor is no longer declared'
        : value === MISSING
          ? 'field no longer exists on the canvas'
          : factor.levels.some((level) => valuesEqual(value, level))
            ? null
            : 'published value does not match any declared level'
      if (reason) discrepancies.push({ nodeId: node.id, nodeLabel, fieldPath, factorName, reason })
    }
  }
  return discrepancies
}
